src/libopensc/card-piv.c

```diff
@@ -467,205 +791,1921 @@ static struct sc_card_driver piv_drv = {
+	}
+
+	if (asn1_sm_response[0].flags & SC_ASN1_PRESENT  /* optional */
+			&& ( ee.value == NULL || ee.len <= 2)) {
+		sc_log(card->ctx,"SM BER-TLV not valid");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	if ((asn1_sm_response[1].flags & SC_ASN1_PRESENT) == 0
+			|| (asn1_sm_response[2].flags & SC_ASN1_PRESENT) == 0) {
+		sc_log(card->ctx,"SM missing status or R-MAC");
+		 r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		 goto err;
+	}
+
+	if (status.len != 2
+			|| status.value == NULL
+			|| rmac8.len != 8
+			|| rmac8.value == NULL) {
+		sc_log(card->ctx,"SM status or R-MAC length invalid");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+	cmac_ctx = CMAC_CTX_new();
+	if (cmac_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+#else
+	mac = EVP_MAC_fetch(NULL, "cmac", NULL);
+	cmac_params[cmac_params_n++] = OSSL_PARAM_construct_utf8_string("cipher", cs->cipher_cbc_name, 0);
+	cmac_params[cmac_params_n] = OSSL_PARAM_construct_end();
+	if (mac == NULL
+			|| (cmac_ctx = EVP_MAC_CTX_new(mac)) == NULL) {
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+#endif
+
+	/*  MCV is first, then BER TLV Encoded Encrypted PIV Data and Status */
+	macdatalen = status.value + status.len - sm_apdu->resp;
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+	if (CMAC_Init(cmac_ctx, priv->sm_session.SKrmac, priv->sm_session.aes_size, (*cs->cipher_cbc)(), NULL) != 1
+			|| CMAC_Update(cmac_ctx, priv->sm_session.R_MCV, MCVlen) != 1
+			|| CMAC_Update(cmac_ctx, sm_apdu->resp, macdatalen) != 1
+			|| CMAC_Final(cmac_ctx, priv->sm_session.R_MCV, &R_MCVlen) != 1) {
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+#else
+	if(!EVP_MAC_init(cmac_ctx, (const unsigned char *)priv->sm_session.SKrmac,
+				priv->sm_session.aes_size, cmac_params)
+			|| !EVP_MAC_update(cmac_ctx, priv->sm_session.R_MCV, MCVlen)
+			|| !EVP_MAC_update(cmac_ctx, sm_apdu->resp, macdatalen)
+			|| !EVP_MAC_final(cmac_ctx, priv->sm_session.R_MCV, &R_MCVlen, MCVlen)) {
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+#endif
+
+	if (memcmp(priv->sm_session.R_MCV, rmac8.value, 8) != 0) {
+		sc_log(card->ctx, "SM 8 bytes of R-MAC do not match received R-MAC");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	
+	ed_ctx = EVP_CIPHER_CTX_new();
+	if (ed_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+
+	/* generate same IV used to encrypt response on card */
+	if (EVP_EncryptInit_ex(ed_ctx, (*cs->cipher_ecb)(), NULL, priv->sm_session.SKenc, zeros) != 1
+			|| EVP_CIPHER_CTX_set_padding(ed_ctx,0) != 1
+			|| EVP_EncryptUpdate(ed_ctx, IV, &outli, priv->sm_session.dec_counter, 16) != 1
+			|| EVP_EncryptFinal_ex(ed_ctx, discard, &outdl) != 1
+			|| outdl != 0) {
+		sc_log(card->ctx,"SM encode failed in OpenSSL");
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	/* some commands to not have response data */
```

Comment:
```suggestion
	/* some commands do not have response data */
```

Reply:
Fixed in 55c41dbf56f0bdda2cc763d60684103b2f50c6c5 as well as a few other spelling and many whitespace, tab and extra blank lines.